lib/datasets/dataset.rb

```diff
@@ -33,9 +33,23 @@ def cache_path
       @cache_path ||= CachePath.new(@metadata.id)
     end
 
-    def download(output_path, url, &block)
-      downloader = Downloader.new(url)
-      downloader.download(output_path, &block)
+    def download(output_path, url, *fallback_urls, &block)
+      urls = [url] + fallback_urls
+      urls.each_with_index do |url, index|
+        downloader = Downloader.new(url)
+        downloader.download(output_path, &block)
+        return
+      rescue Net::HTTPClientException => error
+        if urls.last != url
+          failed_url = url
+          fallback_url = urls[index + 1]
+          message = "#{error.response.code} #{error.response.message}: " +
+                    "fallback: <#{failed_url}> -> <#{fallback_url}>"
+          $stderr.puts(message)
+        else
+          raise
+        end
+      end
```

Comment:
Could you try moving this implementation to `Downloader`?

Reply:
Sure! I'll try it.

Reply:
I’ve moved the implementation to the `Downloader` and extracted the URL parsing and validation logic to a method.

I’ve also updated the pull request description. Could you review it again?